### ISSUE TITLE
can't use v[0-9]+ as label (vstring)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9415]
Migrated from [rt.perl.org#56880](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=56880) (status was 'resolved')

Searchable as RT56880$

**Comment by p5pRT** [comment]
### From @mauke

#### Created by @mauke



% perl \-e 'v0&#8203;: ;' 
syntax error at \-e line 1\, near "v0&#8203;:"
Execution of \-e aborted due to compilation errors\.
% perl \-e 'v0x&#8203;: ;'
%

v\-string parsing gets in the way of statement labels\. Adding any
non\-digit character makes it parse as a bareword/label\. I think this is
a bug because \(with the exception of quoting operators\) perl lets you
use any keywords as labels\. This issue is similar to the \(v123 => 45\)
thing fixed in 5\.8\.1 \(5\.8\.0 parsed it as chr\(123\) instead of 'v123'\)\.
It seems inconsistent to treat labels differently\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

This perlbug was built using Perl v5.8.8 - Wed Mar 22 20:53:12 CET 2006
It is being executed now by  Perl v5.10.0 - Wed Dec 19 19:01:35 CET 2007.

Site configuration information for perl v5.10.0:

Configured by mauke at Wed Dec 19 19:01:35 CET 2007.

Summary of my perl5 (revision 5 version 10 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.6.22-gentoo-r9, archname=i686-linux
    uname='linux nora 2.6.22-gentoo-r9 #3 preempt sat nov 3 22:12:09 cet 2007 i686 amd athlon(tm) 64 processor 3200+ authenticamd gnulinux '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O2 -march=native -fomit-frame-pointer',
    cppflags='-fno-strict-aliasing -pipe -I/usr/local/include -I/usr/include/gdbm'
    ccversion='', gccversion='4.2.2', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.6.1.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.6.1'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -march=native -fomit-frame-pointer -L/usr/local/lib'

Locally applied patches:
    SAFEARGV0 - disable magic open in <ARGV>
    METHQUAL0 - allow qualified method calls without a valid object
    UNWARN0 - disable stupid warnings for print and qw


@INC for perl v5.10.0:
    /home/mauke/usr/local/lib/perl5/5.10.0/i686-linux
    /home/mauke/usr/local/lib/perl5/5.10.0
    /home/mauke/usr/local/lib/perl5/site_perl/5.10.0/i686-linux
    /home/mauke/usr/local/lib/perl5/site_perl/5.10.0
    .


Environment for perl v5.10.0:
    HOME=/home/mauke
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_COLLATE=POSIX
    LD_LIBRARY_PATH=/home/mauke/usr/local/lib
    LOGDIR (unset)
    PATH=/home/mauke/usr/local/bin:/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/i686-pc-linux-gnu/gcc-bin/4.1.2:/opt/sun-jdk-1.4.2.13/bin:/opt/sun-jdk-1.4.2.13/jre/bin:/opt/sun-jdk-1.4.2.13/jre/javaws:/usr/kde/3.5/bin:/usr/qt/3/bin:/usr/games/bin
    PERL_BADLANG (unset)
    PERL_UNICODE=SAL
    SHELL=/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From renee.baecker@smart-websolutions.de

Hi\,

attached is a small patch that allows v\[0\-9\]\+ to be used as a label\.

Cheers\,
Renee

**Comment by p5pRT** [comment]
### From renee.baecker@smart-websolutions.de

<details><summary>version.t.patch</summary>

```diff
--- ../orig/perl-5.10.0/lib/version.t	2007-12-18 11:47:07.000000000 +0100
+++ lib/version.t	2008-10-28 11:09:45.000000000 +0100
@@ -551,7 +551,11 @@
 	is $alpha2, $alpha1, "Don't fall for Data::Dumper's tricks";
     }
 
-
+    # http://rt.perl.org/rt3/Ticket/Display.html?id=56880
+    my $counter;
+    eval '$counter = 0; v23: $counter++; goto v23 unless $counter == 2';
+    is $counter, 2, "Use v[0-9]+ as a label"; 
+       
 }
 
 1;

```
</details>

**Comment by p5pRT** [comment]
### From renee.baecker@smart-websolutions.de

<details><summary>toke.c.patch</summary>

```diff
--- ../orig/perl-5.10.0/toke.c	2007-12-18 11:47:08.000000000 +0100
+++ toke.c	2008-10-28 10:51:12.000000000 +0100
@@ -5106,6 +5106,9 @@
 		s = scan_num(s, &pl_yylval);
 		TERM(THING);
 	    }
+            else if ( *start == ':' ){
+               goto keylookup;
+            }  
 	    /* avoid v123abc() or $h{v1}, allow C<print v10;> */
 	    else if (!isALPHA(*start) && (PL_expect == XTERM
 			|| PL_expect == XREF || PL_expect == XSTATE

```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @rgs

2008/10/29 Renée Bäcker \<renee\.baecker@&#8203;smart\-websolutions\.de>&#8203;:
> Hi\,
>
> attached is a small patch that allows v\[0\-9\]\+ to be used as a label\.

Obvious question \-\- how does it react in the presence of double
colons\, like in v10&#8203;::foo ?
I would probably put the condition with the isALPHA below so the
PL\_expect is tested too\.

\(Also\, I would suggest to send only one patch\, and not one patch per
file modified\, and make them appliable without problems by using the
same number of slashes in the \+\+\+ and \-\-\- lines\. Actually git\-diff and
even better git\-format\-patch do that for you\. But all those rants will
be better put in the P5 wiki and in the perlhack\.pod document\, will
come real soon now\.\.\.\)

> \-\-\- \.\./orig/perl\-5\.10\.0/toke\.c  2007\-12\-18 11&#8203;:47&#8203;:08\.000000000 \+0100
> \+\+\+ toke\.c      2008\-10\-28 10&#8203;:51&#8203;:12\.000000000 \+0100
> @&#8203;@&#8203; \-5106\,6 \+5106\,9 @&#8203;@&#8203;
>                s = scan\_num\(s\, &pl\_yylval\);
>                TERM\(THING\);
>            \}
> \+            else if \( \*start == '&#8203;:' \)\{
> \+               goto keylookup;
> \+            \}
>            /\* avoid v123abc\(\) or $h\{v1\}\, allow C\<print v10;> \*/
>            else if \(\!isALPHA\(\*start\) && \(PL\_expect == XTERM
>                        || PL\_expect == XREF || PL\_expect == XSTATE
>
>



\-\- 
'Do you know what they called a sausage\-in\-a\-bun in Quirm?' said Mr Pin\,
as the two walked away\. \-\-\- 'No?' said Mr Tulip\. \-\-\- 'They called it "le
sausage\-in\-le\-bun"\.'                       \-\- Terry Pratchett\, The Truth

**Comment by p5pRT** [comment]
### From renee.baecker@smart-websolutions.de

Hi Rafael\,

v10&#8203;::foo didn't work in 5\.10\.0&#8203;:

not ok 260 \- Use v\[0\-9\]\+ as a label

\#   Failed test 'Use v\[0\-9\]\+ as a label'
\#   at \.\./\.\./perl\-5\.10\.0/lib/version\.t line 557\.
\#          got&#8203;: '0'
\#     expected&#8203;: '2'
Bareword found where operator expected at \(eval 160\) line 1\, near
"v10&#8203;::foo&#8203;::test2"
Bareword found where operator expected at \(eval 160\) line 1\, near
"v10&#8203;::test"
not ok 261 \- call a function in package v10&#8203;::foo

\#   Failed test 'call a function in package v10&#8203;::foo'
\#   at \.\./\.\./perl\-5\.10\.0/lib/version\.t line 561\.
\#          got&#8203;: ''
\#     expected&#8203;: 'v10&#8203;::foo'

With the attached patch all works fine\. And make test was successful\.

Is the patch as it should look like?

Cheers\,
Renee



Rafael Garcia\-Suarez wrote&#8203;:
> 2008/10/29 Renée Bäcker \<renee\.baecker@&#8203;smart\-websolutions\.de>&#8203;:
>   
>> Hi\,
>>
>> attached is a small patch that allows v\[0\-9\]\+ to be used as a label\.
>>     
>
> Obvious question \-\- how does it react in the presence of double
> colons\, like in v10&#8203;::foo ?
> I would probably put the condition with the isALPHA below so the
> PL\_expect is tested too\.
>
> \(Also\, I would suggest to send only one patch\, and not one patch per
> file modified\, and make them appliable without problems by using the
> same number of slashes in the \+\+\+ and \-\-\- lines\. Actually git\-diff and
> even better git\-format\-patch do that for you\. But all those rants will
> be better put in the P5 wiki and in the perlhack\.pod document\, will
> come real soon now\.\.\.\)
>
>   
>> \-\-\- \.\./orig/perl\-5\.10\.0/toke\.c  2007\-12\-18 11&#8203;:47&#8203;:08\.000000000 \+0100
>> \+\+\+ toke\.c      2008\-10\-28 10&#8203;:51&#8203;:12\.000000000 \+0100
>> @&#8203;@&#8203; \-5106\,6 \+5106\,9 @&#8203;@&#8203;
>>                s = scan\_num\(s\, &pl\_yylval\);
>>                TERM\(THING\);
>>            \}
>> \+            else if \( \*start == '&#8203;:' \)\{
>> \+               goto keylookup;
>> \+            \}
>>            /\* avoid v123abc\(\) or $h\{v1\}\, allow C\<print v10;> \*/
>>            else if \(\!isALPHA\(\*start\) && \(PL\_expect == XTERM
>>                        || PL\_expect == XREF || PL\_expect == XSTATE
>>
>>
>>     
>
>
>
>   


\-\- 


---
renee\.baecker@&#8203;smart\-websolutions\.de

Smart\-Websolutions André Windolph und Renée Bäcker GbR
Maria\-Montessori\-Str\. 13
64584 Biebesheim

USt\.\-ID&#8203;: DE 228 935 695

Telefon&#8203;: 0177\-2416996
XING\-Profil&#8203;: https://www.xing.com/profile/Renee_Baecker

Perl\-Magazin&#8203;: http&#8203;://foo\-magazin\.de
Perl\-Nachrichten&#8203;: http&#8203;://perl\-nachrichten\.de

**Comment by p5pRT** [comment]
### From renee.baecker@smart-websolutions.de

<details><summary>56880.patch</summary>

```diff
--- toke.c.orig	2008-10-30 09:27:25.000000000 +0100
+++ toke.c	2008-10-30 09:06:38.000000000 +0100
@@ -5106,6 +5106,11 @@
 		s = scan_num(s, &pl_yylval);
 		TERM(THING);
 	    }
+            else if ( *start == ':' ){
+                if( start[1] != ':' ){
+                    goto keylookup;
+                }
+            }  
 	    /* avoid v123abc() or $h{v1}, allow C<print v10;> */
 	    else if (!isALPHA(*start) && (PL_expect == XTERM
 			|| PL_expect == XREF || PL_expect == XSTATE
--- lib/version.t.orig	2008-10-30 09:27:39.000000000 +0100
+++ lib/version.t	2008-10-30 09:13:00.000000000 +0100
@@ -551,7 +551,15 @@
 	is $alpha2, $alpha1, "Don't fall for Data::Dumper's tricks";
     }
 
+    # http://rt.perl.org/rt3/Ticket/Display.html?id=56880
+    my $counter = 0;
+    eval 'v23: $counter++; goto v23 unless $counter == 2';
+    is $counter, 2, "Use v[0-9]+ as a label";
 
+    my $output = "";
+    eval "package v10::foo; sub test2 { return 'v10::foo' } package v10; sub test { return v10::foo::test2(); } package main; \$output = v10::test(); "; 
+    is $output, 'v10::foo', 'call a function in package v10::foo';
+       
 }
 
 1;

```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Oct 29 00&#8203;:13&#8203;:47 2008\, renee\.baecker@&#8203;smart\-websolutions\.de wrote&#8203;:
> Hi\,
> 
> attached is a small patch that allows v\[0\-9\]\+ to be used as a label\.
> 
> Cheers\,
> Renee
> 

That patch looks good except for one thing&#8203;:

$ \./miniperl \-le' $\_ = 1 ? v65 : "bar"; print'
A
$ \./miniperl \-le' $\_ = 1 ? v65&#8203;: "bar"; print'
v65

Do you need to check for XTERM or something?

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed by e9d2327\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
